src/smt/term_formula_removal.cpp

```diff
@@ -105,77 +106,110 @@ Node RemoveTermFormulas::run(TNode node, std::vector<Node>& output,
     }
-    Node skolem_app = nodeManager->mkNode( kind::APPLY_UF, skolem_app_c );
-    children.push_back( skolem_app.eqNode( node[1] ) );
-    // axiom defining skolem
-    newAssertion = nodeManager->mkNode( kind::FORALL, children );
   }
 
   // If a Hilbert choice function, witness the choice.
   //   For details on this operator, see
   //   http://planetmath.org/hilbertsvarepsilonoperator.
   if (node.getKind() == kind::CHOICE && !inQuant)
   {
-    // Make the skolem to witness the choice
-    skolem = nodeManager->mkSkolem(
-        "choiceK",
-        nodeType,
-        "a skolem introduced due to term-level Hilbert choice removal");
-
-    Assert(node[0].getNumChildren() == 1);
-
-    // The new assertion is the assumption that the body
-    // of the choice operator holds for the Skolem
-    newAssertion = node[1].substitute(node[0][0], skolem);
+    skolem = getSkolemForNode(node);
+    if (skolem.isNull())
+    {
+      // Make the skolem to witness the choice
+      skolem = nodeManager->mkSkolem(
+          "choiceK",
+          nodeType,
+          "a skolem introduced due to term-level Hilbert choice removal");
+      d_skolem_cache.insert(node, skolem);
+
+      Assert(node[0].getNumChildren() == 1);
+
+      // The new assertion is the assumption that the body
+      // of the choice operator holds for the Skolem
+      newAssertion = node[1].substitute(node[0][0], skolem);
+    }
   }
 
   //if a non-variable Boolean term, replace it
   if(node.getKind()!=kind::BOOLEAN_TERM_VARIABLE && nodeType.isBoolean() && inTerm && !inQuant ){//(!inQuant || !node.hasBoundVar())){
+    skolem = getSkolemForNode(node);
+    if (skolem.isNull())
+    {
+      // Make the skolem to represent the Boolean term
+      // skolem = nodeManager->mkSkolem("termBT", nodeType, "a variable
+      // introduced due to Boolean term removal");
+      skolem = nodeManager->mkBooleanTermVariable();
+      d_skolem_cache.insert(node, skolem);
 
-    // Make the skolem to represent the Boolean term
-    //skolem = nodeManager->mkSkolem("termBT", nodeType, "a variable introduced due to Boolean term removal");
-    skolem = nodeManager->mkBooleanTermVariable();
-
-    // The new assertion
-    newAssertion = skolem.eqNode( node );
+      // The new assertion
+      newAssertion = skolem.eqNode(node);
+    }
   }
 
-  // if we introduced a skolem
+  // if the term should be replaced by a skolem
   if( !skolem.isNull() ){
-    Debug("ite") << "removeITEs(" << node << ") => " << newAssertion << endl;
-
     // Attach the skolem
-    d_iteCache.insert(cacheKey, skolem);
+    d_tfCache.insert(cacheKey, skolem);
 
-    // Remove ITEs from the new assertion, rewrite it and push it to the output
-    newAssertion = run(newAssertion, output, iteSkolemMap, false, false);
+    // if the skolem was introduced in this call
+    if (!newAssertion.isNull())
+    {
+      Debug("ite") << "*** term formula removal introduced " << skolem
```

Comment:
Use more general debug prefix? Since this is not only ITE removal? `Debug("ite")` occurs four times in this file.